### PR TITLE
fixes duplicate collection contents

### PR DESCRIPTION
**Pull request by anjor**
fixes https://github.com/application-research/estuary/issues/699

Test:

```
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   158  100   158    0     0  29867      0 --:--:-- --:--:-- --:--:--  154k
[
  {
    "name": "dir1",
    "type": "directory",
    "size": 0,
    "contId": 0,
    "dir": "/",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-07T23:02:06.151684Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   162  100   162    0     0  47011      0 --:--:-- --:--:-- --:--:--  158k
[
  {
    "name": "dir2",
    "type": "directory",
    "size": 0,
    "contId": 0,
    "dir": "/dir1",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-07T23:02:06.151684Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1/dir2" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   394  100   394    0     0   106k      0 --:--:-- --:--:-- --:--:--  384k
[
  {
    "name": "a",
    "type": "file",
    "size": 13,
    "contId": 85,
    "cid": "bafkreidame73g74ml6273rlrqp5vmii5s3e37nps6guduwlguei3k4ia54",
    "dir": "/dir1/dir2",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-07T23:02:06.151684Z"
  },
  {
    "name": "dir3",
    "type": "directory",
    "size": 0,
    "contId": 0,
    "dir": "/dir1/dir2",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-08T09:35:40.964261Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1/dir2/a" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   231  100   231    0     0  57121      0 --:--:-- --:--:-- --:--:--  225k
[
  {
    "name": "a",
    "type": "file",
    "size": 13,
    "contId": 85,
    "cid": "bafkreidame73g74ml6273rlrqp5vmii5s3e37nps6guduwlguei3k4ia54",
    "dir": "/dir1/dir2/a",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-07T23:02:06.151684Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1/dir2/dir3" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   235  100   235    0     0  58240      0 --:--:-- --:--:-- --:--:--  229k
[
  {
    "name": "a",
    "type": "file",
    "size": 13,
    "contId": 114,
    "cid": "bafkreidame73g74ml6273rlrqp5vmii5s3e37nps6guduwlguei3k4ia54",
    "dir": "/dir1/dir2/dir3",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-08T09:35:40.964261Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1/dir2/dir3/a" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   237  100   237    0     0  62931      0 --:--:-- --:--:-- --:--:--  231k
[
  {
    "name": "a",
    "type": "file",
    "size": 13,
    "contId": 114,
    "cid": "bafkreidame73g74ml6273rlrqp5vmii5s3e37nps6guduwlguei3k4ia54",
    "dir": "/dir1/dir2/dir3/a",
    "coluuid": "b9f372de-0301-4fc3-b70d-53a140b06143",
    "updatedAt": "2022-12-08T09:35:40.964261Z"
  }
]
anjor@seven ~ $ curl -k "http://localhost:3004/collections/$coluuid?dir=/dir1/dir4" -H "Authorization: Bearer $(cat ~/.estuary_local)" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100     5  100     5    0     0   1488      0 --:--:-- --:--:-- --:--:--  5000
null

```

### CODE REVIEW

**Comment by anjor** [comment]
dont merge it yet, i plan to extract the logic into a function and write some unit tests.

File: collections/collections.go [hunk 1]
```diff
@@ -31,6 +34,29 @@ type CollectionRef struct {
...
+	Type      CidType     `json:"type"`
+	Size      int64       `json:"size"`
+	ContID    uint        `json:"contId"`
+	Cid       *util.DbCID `json:"cid,omitempty"`
+	Dir       string      `json:"dir"`
+	ColUuid   string      `json:"coluuid"`
+	UpdatedAt time.Time   `json:"updatedAt"`
+}
+
+const (
```
Comment (en0ma) [inline collections/collections.go:50]: move cidtype to util and reuse where needed

File: collections/collections.go [hunk 2]
```diff
@@ -75,3 +101,99 @@ func GetContentsInPath(coluuid string, path string, db *gorm.DB, u *util.User) (
...
+
+func GetDirectoryContents(c echo.Context, refs []util.ContentWithPath, queryDir, coluuid string) ([]CollectionListResponse, error) {
+	dirs := make(map[string]bool)
+	var result []CollectionListResponse
+	for _, r := range refs {
+		directoryContent, traversedDirectories, err := getDirectoryContent(r, dirs, queryDir, coluuid)
+		dirs = traversedDirectories
+
+		if err != nil {
+			if err == errors.New(ERR_INTERNAL_SERVER) {
```
Comment (en0ma) [inline collections/collections.go:113]: use util.HTTP error

File: collections/collections.go [hunk 3]
```diff
@@ -31,6 +34,29 @@ type CollectionRef struct {
...
+	ColUuid   string      `json:"coluuid"`
+	UpdatedAt time.Time   `json:"updatedAt"`
+}
+
+const (
+	CidTypeDir  CidType = "directory"
+	CidTypeFile CidType = "file"
+)
+
+const (
```
Comment (en0ma) [inline collections/collections.go:55]: no need for this, use util.Http error

File: collections/collections.go (hunk 2 above)
Comment (en0ma) [reply collections/collections.go:113]: also, for 500, just return the error, the error handler will take care of the rest

**Review by en0ma — DISMISSED** [review]
Decent!

**Comment by gmelodie** [comment]
Nice! Thanks for the tests @anjor !

**Pull request merged**